release-notes/swan-lake-beta4-release-note.md

```diff
@@ -34,9 +34,480 @@ If you have not installed Ballerina, then download the [installers](/downloads/#
+
+A function having an optional type that is not a subtype of `error?` as the return type is now expected to explicitly return a value. A warning is emitted when such a function does not explicitly return a value and falls off the end of the function body. 
+
+```ballerina
+function parse(string str) returns int? { // Now results in a warning. 
+    int|error a = int:fromString(str);
+    if a is int {
+        return a;
+    }
+}
+```
+
+#### Bug Fixes and Breaking Changes
+
+- The trailing dot format of the floating point literal has been disallowed to avoid lexical ambiguity.
+
+```ballerina
+// The following are now disallowed.
+decimal d1 = 2.;
+decimal d2 = 2.d;
+decimal d3 = 2.D;
+decimal d4 = 2.e12;
+float f1 = 2.f;
+float f2 = 2.F;
+float f3 = 0x1A.;
+float f4 = 0x1A.p4;
+
+// The following can be used instead.
+decimal d11 = 2.0;
+decimal d12 = 2.0d;
+decimal d13 = 2.0D;
+decimal d14 = 2.0e12;
+float f11 = 2.0f;
+float f12 = 2.0F;
+float f13 = 0x1A.0;
+float f14 = 0x1A.0p4;
+```
+
+- Intervening white spaces have been disallowed in the qualified identifier to avoid a parsing ambiguity between the ternary conditional expression and qualified identifier.
+  
+```ballerina
+import ballerina/io;
+
+public function main() {
+    io:print("Ballerina"); // Valid.
+    io : print("Ballerina"); // compilation error: intervening whitespaces are not allowed in a qualified identifier.
+}
+```
+
+With this, `x ? a : b:c` will now be parsed as `x ? a : (b:c)` since the colon with spaces is interpreted only as part of a conditional expression.
+
+- A bug that resulted in hash collisions not being handled correctly in `table` values has been fixed. 
+
+```ballerina
+import ballerina/io;
+ 
+public function main() {
+    table<record {readonly int? k;}> key(k) t = table [];
+    t.add({k: 0});
+    io:println(t.hasKey(()));
+}
+```
+
+The above code snippet, which previously printed `true` will now print `false`.
+
+- Object type inclusion with an object that has private fields or members has been disallowed.
+
+```ballerina
+class Person {
+    string firstName;
+    string lastName;
+    private string dob;
+
+    function init(string firstName, string lastName, string dob) {
+        self.firstName = firstName;
+        self.lastName = lastName;
+        self.dob = dob;
+    }
+
+    private function getName() returns string => self.firstName + self.lastName;
+}
+ 
+class Employee {
+    *Person; // Will now result in an error.
+    int id;
+
+    function init(string firstName, string lastName, string dob, int id) {
+        self.firstName = firstName;
+        self.lastName = lastName;
+        self.dob = dob;       
+        self.id = id;
+    }
+
+    private function getName() returns string => self.firstName;	
+}
+```
+
+- A bug that resulted in compilation errors not being emitted for invalid `xml` template expressions has been fixed.
+
+```ballerina
+xml x = xml `</>`; // Will now result in an error.
+```
+
```

Comment:
````suggestion

- A bug that resulted in compilation errors not being emitted for duplicate fields written with escape sequences in the mapping constructor has been fixed.
```ballerina
map<any> x = {a\\: 454, "a\\": false}; // Will now result in an error.
```

````